Add IDL files and implementation of CookieChangeEvent

RupinMittal opened the pull request:
#### 0fb388663dbd3ffced8adfc4c1bb722aebf44d30
<pre>
Add IDL files and implementation of CookieChangeEvent
<a href="https://bugs.webkit.org/show_bug.cgi?id=259932">https://bugs.webkit.org/show_bug.cgi?id=259932</a>

Reviewed by Chris Dumez.

CookieChangeEvent and CookieChangeEventInit are needed
for the change events of the CookieStore API. This patch
adds the needed IDL and implementation files.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_arguments.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/idlharness.tentative.https.any-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp: Copied from Source/WebCore/Modules/cookie-store/CookieListItem.idl.
(WebCore::CookieChangeEvent::create):
(WebCore::CookieChangeEvent::CookieChangeEvent):
(WebCore::CookieChangeEvent::eventInterface const):
* Source/WebCore/Modules/cookie-store/CookieChangeEvent.h: Copied from Source/WebCore/Modules/cookie-store/CookieListItem.idl.
* Source/WebCore/Modules/cookie-store/CookieChangeEvent.idl: Copied from Source/WebCore/Modules/cookie-store/CookieListItem.idl.
* Source/WebCore/Modules/cookie-store/CookieChangeEventInit.h: Copied from Source/WebCore/Modules/cookie-store/CookieListItem.idl.
* Source/WebCore/Modules/cookie-store/CookieChangeEventInit.idl: Copied from Source/WebCore/Modules/cookie-store/CookieListItem.idl.
* Source/WebCore/Modules/cookie-store/CookieListItem.h:
* Source/WebCore/Modules/cookie-store/CookieListItem.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.in:

Canonical link: <a href="https://commits.webkit.org/266739@main">https://commits.webkit.org/266739@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2fb3fdb8f3721d03e0dd4631a3d1fcecb2b235a7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14579 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14889 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15234 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16326 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13769 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14715 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17404 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14966 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16427 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14761 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15277 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12379 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17057 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12560 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13148 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20156 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13639 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13314 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16552 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13865 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/11769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13159 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3526 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17494 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13709 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebCore/Modules/cookie-store/CookieListItem.h:
```diff
@@ -63,8 +63,8 @@ struct CookieListItem {
-    bool secure;
-    CookieSameSite sameSite;
+    bool secure { false };
+    CookieSameSite sameSite { CookieSameSite::Strict };
```
I'm not sure that these changed should be made. The IDL files do not specify that we should have default values for `sameSite` and `secure`. But, if we do not have the default values, the tests that are newly passing with this patch will not pass. The reason seems to be that the tests do not specify a value for `sameSite` when creating the `CookieChangeEvent`. The JS bindings default initialize the `CookieListItems` (which don't have a default value for `sameSite`) and since no value for `sameSite` is given, `sameSite` gets some garbage value. This causes us to hit the `ASSERT(static_cast<size_t>(enumerationValue) < std::size(values));` assertion at `JSCookieSameSite.cpp:43`.  At the moment, I've uploaded the patch (as a draft) with the default values so we can see how the tests pass, but I'm not sure what the correct course of action is here. Should we use default values? Or alter the test? Or perhaps something else?

cdumez replied:
Seems fine.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp:
```diff
@@ -0,0 +1,70 @@
+        m_deleted = WTFMove(eventInitDict->deleted);
+    }
+
+    ALWAYS_LOG_WITH_STREAM(stream << "the size is " << m_changed.size());
```
Please drop this logging.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp:
```diff
@@ -0,0 +1,70 @@
+
+CookieChangeEvent::~CookieChangeEvent() = default;
+
+const Vector<CookieListItem>& CookieChangeEvent::changed() const
```
Please inline in header.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp:
```diff
@@ -0,0 +1,70 @@
+    return m_changed;
+}
+
+const Vector<CookieListItem>& CookieChangeEvent::deleted() const
```
ditto.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.h:
```diff
@@ -0,0 +1,55 @@
+    // Event
+    EventInterface eventInterface() const override;
+
+    Vector<CookieListItem> m_changed = { };
```
`= { }` is not needed.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.h:
```diff
@@ -0,0 +1,55 @@
+    EventInterface eventInterface() const override;
+
+    Vector<CookieListItem> m_changed = { };
+    Vector<CookieListItem> m_deleted = { };
```
`= { }` is not needed.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.h:
```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include "CookieChangeEventInit.h"
+#include "CookieListItem.h"
```
Hopefully a forward declaration suffices.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.h:
```diff
@@ -0,0 +1,55 @@
+    const Vector<CookieListItem>& deleted() const;
+
+private:
+    explicit CookieChangeEvent(const AtomString& type, std::optional<CookieChangeEventInit>&&);
```
explicit is not needed since the second parameter is mandatory

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.h:
```diff
@@ -0,0 +1,55 @@
+    explicit CookieChangeEvent(const AtomString& type, std::optional<CookieChangeEventInit>&&);
+
+    // Event
+    EventInterface eventInterface() const override;
```
override -> final

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEventInit.h:
```diff
@@ -0,0 +1,47 @@
+    CookieChangeEventInit() = default;
+
+    CookieChangeEventInit(Vector<CookieListItem>&& changedCookies, Vector<CookieListItem>&& deletedCookies)
+        : EventInit()
```
Not needed.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEventInit.h:
```diff
@@ -0,0 +1,47 @@
+namespace WebCore {
+
+struct CookieChangeEventInit : EventInit {
+    CookieChangeEventInit() = default;
```
Do we really need those 2 constructors?

RupinMittal replied:
I believe we can do without them. Fixed!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.h:
```diff
@@ -0,0 +1,56 @@
+class CookieChangeEvent final : public Event {
+    WTF_MAKE_ISO_ALLOCATED(CookieChangeEvent);
+public:
+    static Ref<CookieChangeEvent> create(const AtomString& type, std::optional<CookieChangeEventInit>);
```
This can just take a `CookieChangeEventInit&&`. The generated bindings code will always pass a `CookieChangeEventInit` to you, since there is a default parameter value in the IDL.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp:
```diff
@@ -0,0 +1,58 @@
+    : Event(type, Event::CanBubble::No, Event::IsCancelable::No)
+{
+    if (eventInitDict) {
+        m_changed = WTFMove(eventInitDict->changed);
```
Would be nice to move to the initializer list.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp:
```diff
@@ -0,0 +1,58 @@
+{
+    if (eventInitDict) {
+        m_changed = WTFMove(eventInitDict->changed);
+        m_deleted = WTFMove(eventInitDict->deleted);
```
ditto.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.idl:
```diff
@@ -0,0 +1,35 @@
+    Exposed=Window,
+    SecureContext
+] interface CookieChangeEvent : Event {
+  constructor([AtomString] DOMString type, optional CookieChangeEventInit eventInitDict = {});
```
Indentation issue, should be 4 space here and below.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEventInit.idl:
```diff
@@ -0,0 +1,33 @@
+typedef sequence<CookieListItem> CookieList;
+
+dictionary CookieChangeEventInit : EventInit {
+  CookieList changed;
```
Indentation is 4 spaces in WebKit.

cdumez reviewed (approved):
LGTM with fixes

cdumez commented on Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp:
```diff
@@ -0,0 +1,55 @@
+}
+
+CookieChangeEvent::CookieChangeEvent(const AtomString& type, CookieChangeEventInit&& eventInitDict)
+    : Event(type, Event::CanBubble::No, Event::IsCancelable::No)
```
This is actually incorrect, you're failing to pass init to the Event() base constructor. `eventInitDict` is what should control bubbles / cancellable via EventInit.

cdumez replied:
See `BlobEvent` as example.

RupinMittal replied:
I'm not entirely sure I understand the issue. For example, `ApplePayCancelEvent` does what I have done above. Although, it does make sense that the whoever is creating the `CookieChangeEvent` should have control over canBubbles/isCancelable. I can add two fields to `CookieChangeEventInit' like so:

```
struct CookieChangeEventInit : EventInit {
    Vector<CookieListItem> changed;
    Vector<CookieListItem> deleted;
    Event::CanBubble canBubble;
    Event::IsCancelable isCancelable;
};
```
  
and then in `CookieChangeEvent::create` I can change it to:
```
CookieChangeEvent::CookieChangeEvent(const AtomString& type, CookieChangeEventInit&& eventInitDict)
    : Event(type, eventInitDict.canBubble, eventInitDict.isCancelable)
    , m_changed(WTFMove(eventInitDict.changed))
    , m_deleted(WTFMove(eventInitDict.deleted))
{ }
```

Would this be preferable?

cdumez replied:
You don’t need to add data members, they are already there.. you inherit them from EventInit.

cdumez replied:
Also Apple Pay code is not necessarily a good reference.

cdumez reviewed (approved):
LGTM

RupinMittal committed:
Add IDL files and implementation of CookieChangeEvent
https://bugs.webkit.org/show_bug.cgi?id=259932

Reviewed by Chris Dumez.

CookieChangeEvent and CookieChangeEventInit are needed
for the change events of the CookieStore API. This patch
adds the needed IDL and implementation files.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_event_arguments.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/idlharness.tentative.https.any-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/cookie-store/CookieChangeEvent.cpp: Copied from Source/WebCore/Modules/cookie-store/CookieListItem.idl.
(WebCore::CookieChangeEvent::create):
(WebCore::CookieChangeEvent::CookieChangeEvent):
(WebCore::CookieChangeEvent::eventInterface const):
* Source/WebCore/Modules/cookie-store/CookieChangeEvent.h: Copied from Source/WebCore/Modules/cookie-store/CookieListItem.idl.
* Source/WebCore/Modules/cookie-store/CookieChangeEvent.idl: Copied from Source/WebCore/Modules/cookie-store/CookieListItem.idl.
* Source/WebCore/Modules/cookie-store/CookieChangeEventInit.h: Copied from Source/WebCore/Modules/cookie-store/CookieListItem.idl.
* Source/WebCore/Modules/cookie-store/CookieChangeEventInit.idl: Copied from Source/WebCore/Modules/cookie-store/CookieListItem.idl.
* Source/WebCore/Modules/cookie-store/CookieListItem.h:
* Source/WebCore/Modules/cookie-store/CookieListItem.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.in:

Canonical link: https://commits.webkit.org/266739@main
